### PR TITLE
Shape inference for onnx.STFT

### PR DESCRIPTION
**Pull request by abhishekbisht1429**
- Implemented shape inference for onnx.STFT

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps/Math/STFT.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===----------------- STFT.cpp - Lowering STFT Op
+//--------------------------===//
+//
+// Copyright 2019-2024 The IBM Research Authors.
```
Comment (jorickert) [inline src/Dialect/ONNX/ONNXOps/Math/STFT.cpp]:
```suggestion
// Copyright 2025 AMD
```

File: src/Dialect/ONNX/ONNXOps/Math/STFT.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===----------------- STFT.cpp - Lowering STFT Op
+//--------------------------===//
```
Comment (ttjost) [inline src/Dialect/ONNX/ONNXOps/Math/STFT.cpp]: NIT: too many characters in the line above.

File: src/Dialect/ONNX/ONNXOps/Math/STFT.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+//
+// This file provides definition of ONNX dialect STFT operation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Dialect/ONNX/DialectBuilder.hpp"
+#include "src/Dialect/ONNX/ONNXOps/OpHelper.hpp"
+
+using namespace mlir;
+using namespace mlir::OpTrait::util;
```
Comment (jorickert) [inline src/Dialect/ONNX/ONNXOps/Math/STFT.cpp]:
```suggestion

```
Not needed namespace

File: src/Dialect/ONNX/ONNXOps/Math/STFT.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+//
+// Copyright 2019-2024 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file provides definition of ONNX dialect STFT operation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Dialect/ONNX/DialectBuilder.hpp"
```
Comment (jorickert) [inline src/Dialect/ONNX/ONNXOps/Math/STFT.cpp]:
```suggestion

```
Unused include

File: src/Dialect/ONNX/ONNXOps/Math/STFT.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXSTFTOp::inferShapes(
+    std::function<void(Region &)> doShapeInference) {
+  // Cannot infer the output shape if the operands shape isn't known yet.
+  if (!hasShapeAndRank(getOperation()))
+    return success();
+
+  Type elementType =
+      mlir::cast<ShapedType>(getSignal().getType()).getElementType();
```
Comment (jorickert) [inline src/Dialect/ONNX/ONNXOps/Math/STFT.cpp:79]: This cast is unsafe, it could be that the operation is shaped, but the signal is not.

Mabye use the operation return element type?

File: src/Dialect/ONNX/ONNXOps/Math/STFT.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+//===----------------------------------------------------------------------===//
+// Support
+//===----------------------------------------------------------------------===//
+
+namespace onnx_mlir {
+
+template <>
+LogicalResult ONNXGenericSTFTOpShapeHelper<ONNXSTFTOp>::computeShape() {
+  typename ONNXSTFTOp::Adaptor operandAdaptor(
```
Comment (jorickert) [inline src/Dialect/ONNX/ONNXOps/Math/STFT.cpp]:
```suggestion
  ONNXSTFTOpAdaptor operandAdaptor(
```

I do not think the typename is needed here

File: src/Dialect/ONNX/ONNXOps/Math/STFT.cpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+      operands, op->getAttrDictionary());
+  auto signal = operandAdaptor.getSignal();
+  if (!hasShapeAndRank(signal))
+    return failure();
+  auto oneSided = operandAdaptor.getOnesided() == 1;
+
+  auto frameLengthIE =
+      createIE->getIntAsSymbol(operandAdaptor.getFrameLength());
```
Comment (ttjost) [inline src/Dialect/ONNX/ONNXOps/Math/STFT.cpp]: The spec https://onnx.ai/onnx/operators/onnx__STFT.html says that `frame_length` and `window` can be optional, and according to https://github.com/onnx/onnx/blob/c12bcfe2637b64d1f21ba2f9b081039812a0cb3d/onnx/defs/math/defs.cc#L3496 , there needs to be at least one of them defined.
Could you please make sure to adapt the code to follow the same implementation as in the link?

Comment (ttjost) [reply src/Dialect/ONNX/ONNXOps/Math/STFT.cpp]: You should, therefore, check if those are optional, and make sure to use the same implementation as the link. Almost everything is done, except these corner cases 😄

File: test/mlir/onnx/onnx_shape_inference.mlir [hunk 8]
```diff
@@ -4070,4 +4070,76 @@ func.func @test_grid_sample_dim_shape3(%arg0: tensor<?x?x?x?xf32>, %arg1: tensor
...
-}
\ No newline at end of file
+}
+
+// -----
+
+// Test STFT
+// Test case 1, with static shapes and onesided = 1
+func.func @test_stft_static_1(%arg0: tensor<1x32x1xf32>) -> tensor<*xf32> {
+  %arg1 = onnx.Constant dense<1> : tensor<i64>
```
Comment (ttjost) [inline test/mlir/onnx/onnx_shape_inference.mlir]: would it be possible to rename these to their correct input names? THat is, `frame_step`, `window`, and `frame_length`.

File: test/mlir/onnx/onnx_shape_inference.mlir [hunk 9]
```diff
@@ -4070,4 +4070,76 @@ func.func @test_grid_sample_dim_shape3(%arg0: tensor<?x?x?x?xf32>, %arg1: tensor
...
+  %arg2 = onnx.Constant dense<[0.5, 0.5, 0.5, 0.5]> : tensor<4xf32>
+  %arg3 = onnx.Constant dense<4> : tensor<i64>
+  %0 = "onnx.STFT"(%arg0, %arg1, %arg2, %arg3) : (tensor<1x32x1xf32>, tensor<i64>, tensor<4xf32>, tensor<i64>) -> tensor<*xf32>
+  return %0 : tensor<*xf32>
+}
+// mlir2FileCheck.py
```
Comment (ttjost) [inline test/mlir/onnx/onnx_shape_inference.mlir]:
```suggestion
```

File: test/mlir/onnx/onnx_shape_inference.mlir [hunk 10]
```diff
@@ -4070,4 +4070,76 @@ func.func @test_grid_sample_dim_shape3(%arg0: tensor<?x?x?x?xf32>, %arg1: tensor
...
+
+// Test case 2, with static shapes and onesided = 1
+func.func @test_stft_static_2(%arg0: tensor<1x32x1xf32>) -> tensor<*xf32> {
+  %arg1 = onnx.Constant dense<3> : tensor<i64>
+  %arg2 = onnx.Constant dense<[0.5, 0.5, 0.5, 0.5]> : tensor<4xf32>
+  %arg3 = onnx.Constant dense<4> : tensor<i64>
+  %0 = "onnx.STFT"(%arg0, %arg1, %arg2, %arg3) : (tensor<1x32x1xf32>, tensor<i64>, tensor<4xf32>, tensor<i64>) -> tensor<*xf32>
+  return %0 : tensor<*xf32>
+}
+// mlir2FileCheck.py
```
Comment (ttjost) [inline test/mlir/onnx/onnx_shape_inference.mlir]:
```suggestion
```

File: test/mlir/onnx/onnx_shape_inference.mlir [hunk 11]
```diff
@@ -4070,4 +4070,76 @@ func.func @test_grid_sample_dim_shape3(%arg0: tensor<?x?x?x?xf32>, %arg1: tensor
...
+
+// Test case 3, with static shape and onesided = 0
+func.func @test_stft_static_onesided_0(%arg0: tensor<1x32x1xf32>) -> tensor<*xf32> {
+  %arg1 = onnx.Constant dense<3> : tensor<i64>
+  %arg2 = onnx.Constant dense<[0.5, 0.5, 0.5, 0.5]> : tensor<4xf32>
+  %arg3 = onnx.Constant dense<4> : tensor<i64>
+  %0 = "onnx.STFT"(%arg0, %arg1, %arg2, %arg3) {onesided = 0 : si64} : (tensor<1x32x1xf32>, tensor<i64>, tensor<4xf32>, tensor<i64>) -> tensor<*xf32>
+  return %0 : tensor<*xf32>
+}
+// mlir2FileCheck.py
```
Comment (ttjost) [inline test/mlir/onnx/onnx_shape_inference.mlir]:
```suggestion
```

File: test/mlir/onnx/onnx_shape_inference.mlir [hunk 12]
```diff
@@ -4070,4 +4070,76 @@ func.func @test_grid_sample_dim_shape3(%arg0: tensor<?x?x?x?xf32>, %arg1: tensor
...
+// CHECK-SAME: : (tensor<1x32x1xf32>, tensor<i64>, tensor<4xf32>, tensor<i64>) -> tensor<1x10x4x2xf32>
+// CHECK:           return [[VAR_3_]] : tensor<1x10x4x2xf32>
+
+
+// Test case 4, with dynamic shape and onesided = 0
+func.func @test_stft_dyn(%arg0: tensor<1x8x1xf32>, %arg1: tensor<i64>, %arg2: tensor<4xf32>, %arg3: tensor<i64>) -> tensor<*xf32> {
+  %0 = "onnx.STFT"(%arg0, %arg1, %arg2, %arg3) : (tensor<1x8x1xf32>, tensor<i64>, tensor<4xf32>, tensor<i64>) -> tensor<*xf32>
+  return %0 : tensor<*xf32>
+}
+// mlir2FileCheck.py
```
Comment (ttjost) [inline test/mlir/onnx/onnx_shape_inference.mlir]:
```suggestion
```

File: test/mlir/onnx/onnx_shape_inference.mlir [hunk 13]
```diff
@@ -4070,4 +4070,76 @@ func.func @test_grid_sample_dim_shape3(%arg0: tensor<?x?x?x?xf32>, %arg1: tensor
 // CHECK:           return [[GRID]] : tensor<?x?x10x20xf32>
 // CHECK:         }
   return %0 : tensor<*xf32>
-}
\ No newline at end of file
+}
+
+// -----
+
```
Comment (ttjost) [inline test/mlir/onnx/onnx_shape_inference.mlir:4076]: Could you also add test cases where `frame_length` and `windows` are optional? A reminder, at least one of them needs to be define and if both are, they need to respect what the implementation https://github.com/onnx/onnx/blob/c12bcfe2637b64d1f21ba2f9b081039812a0cb3d/onnx/defs/math/defs.cc#L3496 says.

**Review by ttjost — COMMENTED** [review]
It is looking really good, we just need to make sure it work for all cases.

**Review by ttjost — APPROVED** [review]
Approving as we discussed about implementing the corner cases. Thanks!

**Review by jorickert — APPROVED** [review]

**Review by ttjost — APPROVED** [review]
